File: src/pos_app/lib/features/shifts/models/shift.dart
import '../../../core/models/money.dart';
import '../../tickets/models/enums.dart';

/// CashMovementType as Sales.Domain numbers it: PayIn=0, PayOut=1.
enum CashMovementType {
  payIn(0, 'PayIn'),
  payOut(1, 'PayOut');

  final int value;
  final String name_;

  const CashMovementType(this.value, this.name_);

  static CashMovementType fromName(String? name) =>
      name == CashMovementType.payOut.name_ ? CashMovementType.payOut : CashMovementType.payIn;
}

/// What a movement was for, as Sales.Domain numbers it. Wage and Advance
/// name an employee and reach Payroll; Supplier, Expense and Partner reach
/// Finance; Other is just a reason.
enum CashMovementKind {
  other(0),
  supplier(1),
  wage(2),
  advance(3),
  expense(4),
  partner(5);

  final int value;

  const CashMovementKind(this.value);

  /// What this kind needs the cashier to pick, if anything
  MovementPick? get picks => switch (this) {
        CashMovementKind.wage || CashMovementKind.advance => MovementPick.employee,
        CashMovementKind.supplier => MovementPick.supplier,
        CashMovementKind.partner => MovementPick.partner,
        CashMovementKind.expense => MovementPick.category,
        CashMovementKind.other => null,
      };

  /// The kinds a pay-out offers, and the kinds a pay-in offers
  static const List<CashMovementKind> forPayOut = [supplier, wage, advance, expense, partner, other];
  static const List<CashMovementKind> forPayIn = [partner, other];
}

/// The list a kind's picker shows
enum MovementPick { employee, supplier, partner, category }

/// Cash put into or taken out of the drawer mid-shift, with its reason
/// (Sales `CashMovementView`).
class CashMovementView {
  final CashMovementType type;
  final double amount;
  final String reason;
  final String recordedBy;
  final DateTime? recordedAt;

  const CashMovementView({
    required this.type,
    required this.amount,
    required this.reason,
    required this.recordedBy,
    this.recordedAt,
  });

  bool get isOut => type == CashMovementType.payOut;

  factory CashMovementView.fromJson(Map<String, dynamic> json) => CashMovementView(
        type: CashMovementType.fromName(json['type'] as String?),
        amount: toNumber(json['amount']),
        reason: json['reason'] as String? ?? '',
        recordedBy: json['recordedBy'] as String? ?? '',
        recordedAt: json['recordedAt'] == null ? null : DateTime.tryParse(json['recordedAt'] as String),
      );
}

/// What one tender took over the shift (Sales `TenderTotal`).
class TenderTotal {
  final PaymentTender tender;
  final double amount;
  final int count;

  const TenderTotal({required this.tender, required this.amount, required this.count});

  factory TenderTotal.fromJson(Map<String, dynamic> json) => TenderTotal(
        tender: PaymentTender.fromName(json['tender'] as String?),
        amount: toNumber(json['amount']),
        count: toInt(json['count']),
      );
}

/// A drawer shift (Sales `ShiftView`): the live X while open, the frozen Z
/// once closed. Money figures are the server's.
class ShiftView {
  final int id;
  final String status;
  final DateTime? openedAt;
  final String? openedBy;
  final double openingFloat;
  final DateTime? closedAt;
  final String? closedBy;
  final double? closingCount;
  final double? expectedCash;
  final double? overShort;
  final List<CashMovementView> movements;
  final int ticketsSettled;
  final double salesTotal;
  final List<TenderTotal> tenderTotals;
  final double changeGiven;
  final double refundsTotal;

  /// Bill, line and loyalty discounts on the shift's tickets, as a positive number
  final double discounts;
  final double cashRefunds;
  final double payInsTotal;
  final double payOutsTotal;

  /// Money taken against tabs during the shift, by tender — beside the
  /// sales, never inside them
  final List<TenderTotal> tabPaymentTenderTotals;
  final double expectedInDrawer;

  const ShiftView({
    required this.id,
    this.status = 'Open',
    this.openedAt,
    this.openedBy,
    this.openingFloat = 0,
    this.closedAt,
    this.closedBy,
    this.closingCount,
    this.expectedCash,
    this.overShort,
    this.movements = const [],
    this.ticketsSettled = 0,
    this.salesTotal = 0,
    this.tenderTotals = const [],
    this.changeGiven = 0,
    this.refundsTotal = 0,
    this.discounts = 0,
    this.cashRefunds = 0,
    this.payInsTotal = 0,
    this.payOutsTotal = 0,
    this.tabPaymentTenderTotals = const [],
    this.expectedInDrawer = 0,
  });

  bool get isClosed => status == 'Closed';

  /// What the drawer should have held at close: the frozen figure when
  /// closed, the live one otherwise
  double get expected => expectedCash ?? expectedInDrawer;

  factory ShiftView.fromJson(Map<String, dynamic> json) {
    DateTime? date(String key) => json[key] == null ? null : DateTime.tryParse(json[key] as String);
    double? optional(String key) => json[key] == null ? null : toNumber(json[key]);
    return ShiftView(
      id: toInt(json['id']),
      status: json['status'] as String? ?? 'Open',
      openedAt: date('openedAt'),
      openedBy: json['openedBy'] as String?,
      openingFloat: toNumber(json['openingFloat']),
      closedAt: date('closedAt'),
      closedBy: json['closedBy'] as String?,
      closingCount: optional('closingCount'),
      expectedCash: optional('expectedCash'),
      overShort: optional('overShort'),
      movements: ((json['movements'] as List<dynamic>?) ?? [])
          .map((e) => CashMovementView.fromJson(e as Map<String, dynamic>))
          .toList(),
      ticketsSettled: toInt(json['ticketsSettled']),
      salesTotal: toNumber(json['salesTotal']),
      tenderTotals: ((json['tenderTotals'] as List<dynamic>?) ?? [])
          .map((e) => TenderTotal.fromJson(e as Map<String, dynamic>))
          .toList(),
      changeGiven: toNumber(json['changeGiven']),
      refundsTotal: toNumber(json['refundsTotal']),
      discounts: toNumber(json['discounts']),
      cashRefunds: toNumber(json['cashRefunds']),
      payInsTotal: toNumber(json['payInsTotal']),
      payOutsTotal: toNumber(json['payOutsTotal']),
      tabPaymentTenderTotals: ((json['tabPaymentTenderTotals'] as List<dynamic>?) ?? [])
          .map((e) => TenderTotal.fromJson(e as Map<String, dynamic>))
          .toList(),
      expectedInDrawer: toNumber(json['expectedInDrawer']),
    );
  }
}

/// `POST /api/shifts/{id}/movements`. The kind says what the money was
/// for; the id and name that go with it are whichever the kind picks
/// (see [CashMovementKind.picks]) — the name travels too so the Z report
/// reads without a lookup.
class CashMovementRequest {
  final CashMovementType type;
  final double amount;
  final String reason;
  final CashMovementKind kind;
  final int? employeeId;
  final String? employeeName;
  final int? supplierId;
  final String? supplierName;
  final int? partnerId;
  final String? partnerName;
  final int? categoryId;

  const CashMovementRequest({
    required this.type,
    required this.amount,
    required this.reason,
    this.kind = CashMovementKind.other,
    this.employeeId,
    this.employeeName,
    this.supplierId,
    this.supplierName,
    this.partnerId,
    this.partnerName,
    this.categoryId,
  });

  Map<String, dynamic> toJson() => {
        'type': type.value,
        'amount': amount,
        'reason': reason,
        'kind': kind.value,
        'employeeId': employeeId,
        'employeeName': employeeName,
        'supplierId': supplierId,
        'supplierName': supplierName,
        'partnerId': partnerId,
        'partnerName': partnerName,
        'categoryId': categoryId,
      };
}

File: src/Sales.Domain/AggregatesModel/ShiftAggregate/Shift.cs
#nullable enable
using Ninja.Sales.Domain.Events;

namespace Ninja.Sales.Domain.AggregatesModel.ShiftAggregate;

/// <summary>
/// Shift aggregate root — one cashier's custody of the drawer: opened with a
/// counted float, fed by cash sales and pay-ins/outs, closed with a counted
/// total. The over/short against the expectation is the whole point: it is
/// computed once at close, from numbers the caller assembled, and frozen.
///
/// A shift never blocks selling — a settle with no open shift simply goes
/// unattributed rather than stopping service (docs/pos-plan.md Q4).
/// </summary>
public class Shift : Entity, IAggregateRoot
{
    public int BranchId { get; private set; }

    public ShiftStatus Status { get; private set; }

    public DateTime OpenedAt { get; private set; }

    public string OpenedBy { get; private set; } = string.Empty;

    /// <summary>The cashier's subject id, so Payroll can mark their attendance; null on shifts from before it was recorded.</summary>
    public string? OpenedByUserId { get; private set; }

    /// <summary>Cash counted into the drawer at open.</summary>
    public decimal OpeningFloat { get; private set; }

    public DateTime? ClosedAt { get; private set; }

    public string? ClosedBy { get; private set; }

    /// <summary>Cash counted in the drawer at close.</summary>
    public decimal? ClosingCount { get; private set; }

    /// <summary>
    /// What the drawer should have held at close:
    /// float + cash sales − change given − cash refunds + cash tab payments
    /// + pay-ins − pay-outs. Frozen at close.
    /// </summary>
    public decimal? ExpectedCash { get; private set; }

    /// <summary>Counted minus expected: positive = over, negative = short.</summary>
    public decimal? OverShort { get; private set; }

    private readonly List<CashMovement> _movements = [];
    public IReadOnlyCollection<CashMovement> Movements => _movements.AsReadOnly();

    public decimal GetPayInsTotal() => _movements.Where(m => m.Type == CashMovementType.PayIn).Sum(m => m.Amount);

    public decimal GetPayOutsTotal() => _movements.Where(m => m.Type == CashMovementType.PayOut).Sum(m => m.Amount);

    protected Shift() { }

    public Shift(int branchId, decimal openingFloat, string openedBy, string? openedByUserId = null)
    {
        if (openingFloat < 0)
            throw new SalesDomainException("The opening float cannot be negative");

        if (string.IsNullOrWhiteSpace(openedBy))
            throw new SalesDomainException("A shift needs the cashier opening it");

        BranchId = branchId;
        Status = ShiftStatus.Open;
        OpeningFloat = openingFloat;
        OpenedAt = DateTime.UtcNow;
        OpenedBy = openedBy;
        OpenedByUserId = string.IsNullOrWhiteSpace(openedByUserId) ? null : openedByUserId;

        // The branch opens for business with the drawer: Tenant.API turns
        // the ordering and reservation flags on off this event
        AddDomainEvent(new ShiftOpenedDomainEvent(this));
    }

    public void AddMovement(CashMovementType type, decimal amount, string reason, string recordedBy,
        CashMovementKind kind = CashMovementKind.Other, int? employeeId = null, string? employeeName = null,
        int? supplierId = null, string? supplierName = null, int? partnerId = null, string? partnerName = null, int? categoryId = null)
    {
        EnsureOpen();

        var movement = new CashMovement(type, amount, reason, recordedBy, kind, employeeId, employeeName,
            supplierId, supplierName, partnerId, partnerName, categoryId);
        _movements.Add(movement);

        // Money handed to staff is Payroll's to account for, money for a
        // supplier, an expense or a partner is Finance's: either goes out
        // through the outbox with the movement
        if (movement.IsStaffPayOut)
            AddDomainEvent(new CashPaidOutToStaffDomainEvent(this, movement));
        else if (movement.IsFinanceMovement)
            AddDomainEvent(new CashMovedDomainEvent(this, movement));
    }

    /// <summary>
    /// Count the drawer and freeze the shift. The cash-sales and change
    /// figures come from the tickets stamped with this shift — the caller
    /// aggregates them; the shift owns the arithmetic and the verdict.
    /// </summary>
    public void Close(decimal closingCount, decimal cashPayments, decimal changeGiven, string closedBy, decimal cashRefunds = 0, decimal cashTabPayments = 0)
    {
        EnsureOpen();

        if (closingCount < 0)
            throw new SalesDomainException("The closing count cannot be negative");

        if (string.IsNullOrWhiteSpace(closedBy))
            throw new SalesDomainException("A shift needs the cashier closing it");

        // Cash refunds left the drawer the way change did; cash taken
        // against a tab went in the way a sale did
        ExpectedCash = OpeningFloat + cashPayments - changeGiven - cashRefunds + cashTabPayments + GetPayInsTotal() - GetPayOutsTotal();
        ClosingCount = closingCount;
        OverShort = closingCount - ExpectedCash;
        ClosedBy = closedBy;
        ClosedAt = DateTime.UtcNow;
        Status = ShiftStatus.Closed;

        // ...and closes with it: both flags go off again
        AddDomainEvent(new ShiftClosedDomainEvent(this));
    }

    private void EnsureOpen()
    {
        if (Status != ShiftStatus.Open)
            throw new SalesDomainException($"Shift {Id} is closed and cannot change.");
    }
}
